### ISSUE TITLE
@BeforeGroups called twice

### ISSUE THREAD

**Issue by juherr** [cbeust/testng#1108]
```java
public class testNG {

    @BeforeGroups(groups = {"smoketests", "functionaltests"})
    public void before() {
        System.out.println("Before Groups");
    }

    @Test(groups = {"smoketests", "functionaltests"})
    public void test() {
        System.out.println("Test");
    }

    @AfterGroups(groups = {"smoketests", "functionaltests"})
    public void after() {
        System.out.println("After Groups");
    }
}
```

Will display:

```
Before Groups
Before Groups
Test
After Groups
```

From: http://stackoverflow.com/q/38796087/4234729

**Comment by krmahadevan** [comment]
@juherr I have tried fixing this issue. Please help review.

**Comment by juherr** [comment]
@krmahadevan Done

**Comment by krmahadevan** [comment]
@juherr Thanks for helping with the review. I have fixed all the review comments.

**Comment by AdityaGaurav** [comment]
@krmahadevan I'm getting the same issue with the 6.11 version.

```
<suite name="Test Suite">
    <test name="GroupTest">
          <groups>
              <run>
                <include name="sanity"/>
              </run>
          </groups>
          <classes>
              <class name="SampleTest"/>
          </classes>
    </test>
</suite>
```

Below is the test class:

```
public class SampleTest {
    @BeforeGroups(groups = {"sanity","regression"})
    void beforeGroup(){
        System.out.println("Before Group");
    }

    @AfterGroups(groups = {"sanity","regression"})
    void afterGroups(){
        System.out.println("After Groups");
    }

    @Test(groups = {"sanity"})
    void m1(){
        System.out.println("m1");
    }

    @Test(groups = {"sanity","regression"})
    void m3(){
        System.out.println("m3");
    }
    @Test(groups = {"sanity"})
    void m4(){
        System.out.println("m4");
    }
    @Test(groups = {"regression"})
    void m5(){
        System.out.println("m5");
    }
    @Test
    void m6(){
        System.out.println(Sample.class.getName());
    }
}   
```
So when I am runnting the testng.xml, getting the following result:

```
Before Group
m1
m2
Before Group
m3
After Groups
m4
After Groups
```
Following is the result when I'm changing **group name** to **regression** in the suite.

```
Before Group
m3
After Groups
m5
After Groups
```
@BefroreGroups and @Aftergroups get executed for both the group entries.

**Comment by krmahadevan** [comment]
@AdityaGaurav - The latest released version of TestNG is `6.14.3`. Please retry using that. If you are still seeing the issue, please file a new issue with a standalone reproducible test
